sympy/simplify/cse_main.py

```diff
@@ -86,6 +86,73 @@ def cse_separate(r, e):
     e = d[False]
     return [reps_toposort(r), e]
 
+
+def cse_minimize_memory(r, e):
```

Comment:
What was the decision about rather adding a flag to cse to reuse variable names? 

Either way, I'm not a fan of this function name. Something like `delete_variables` would be more descriptive.

Reply:
But it's not deleting any variables; it is just identifying some as not being needed past a certain element in the returned list of replacements.

Reuse (or minimal) is not as trivial to implement but that could be done, too. I am not sure that reusing will actually release all variables that are not needed, however. If I am right then it, alone, would not be a solution.

Maybe this function should be called `cse_minimize_variables` and do both compaction and identification of redundancy.

Reply:
This seems like a problem that could be solved with the `connected_components` functions...